test/engine.test.js

```diff
@@ -832,4 +833,83 @@ describe("Pipeline Engine tests", function () {
         assert.ok(result.renditionErrors);
         assert.strictEqual(result.renditionErrors[0].message,  "data:image/a/c;base64,xxx must be a valid https url or datauri");
     });
+
+    it("Should try to download before refinePlan with input url no source name nor path", async function () {
+        let downloadRan = false;
+        mockRequire('../lib/storage/http', {
+            download() { 
+                downloadRan = true;
+                console.log(`Fake download success`);
+            }
+        });
+        mockRequire('../lib/metadata', MockMetadata);
+        mockRequire.reRequire('../lib/storage');
+        mockRequire.reRequire('../lib/metadata');
+        const Engine = mockRequire.reRequire('../lib/engine');
+        const pipeline = new Engine();
+
+        const goodTransformer = new GoodTransformer();
+        pipeline.registerTransformer(goodTransformer);
+        pipeline.registerTransformer(new BadTransformer());
+
+        const input = {
+            type: 'image/tiff',
+            url: 'https://example.com/fakeEarth.tiff'
+        };
+        const output = {
+            type: 'image/gif'
+        };
+
+        const plan = new Plan();
+        // TODO: should happen inside transformer
+        await pipeline.refinePlan(plan, input, output);
+        assertPlan(plan, "[start] -> { goodTransformer* }", [{
+            name: "goodTransformer",
+            output: output,
+            input: input
+        }]);
+
+        await pipeline.run(plan);
+        assert.ok(downloadRan);
+    });
+
+    it("Should download before refinePlan with input url (refinePlan)", async function () {
+        let downloadRan = false;
+        mockRequire('../lib/storage/http', {
+            download() { 
+                downloadRan = true;
+                console.log(`Fake download success`);
+            }
+        });
+        mockRequire('../lib/metadata', MockMetadata);
+        mockRequire.reRequire('../lib/storage');
+        mockRequire.reRequire('../lib/metadata');
+        const Engine = mockRequire.reRequire('../lib/engine');
+        const pipeline = new Engine();
+
+        const goodTransformer = new GoodTransformer();
+        pipeline.registerTransformer(goodTransformer);
+        pipeline.registerTransformer(new BadTransformer());
+
+        const input = {
+            type: 'image/tiff',
+            url: 'https://example.com/fakeEarth.tiff',
+            name: "fakeEarth.tiff"
+        };
+        const output = {
+            type: 'image/gif'
+        };
+
+        const plan = new Plan();
+        // TODO: should happen inside transformer
+        await pipeline.refinePlan(plan, input, output);
+        assertPlan(plan, "[start] -> { goodTransformer* }", [{
+            name: "goodTransformer",
+            output: output,
+            input: input
+        }]);
+
+        await pipeline.run(plan);
+        assert.ok(downloadRan);
```

Comment:
move line 83 (assert.ok(downloadRan);) to make sure it ran in refinePlan instead of pipeline.run

Reply:
done